File: 5. Longest Palindromic Substring/longestPalindrome.py
class Solution:
    def longestPalindrome(self, s):
        # write your code here
        if not s:
            return s
            
        ans = ""
        for mid in range(len(s)):
            longest = self.getPalindrome(s, mid, mid)
            ans = longest if len(longest) > len(ans) else ans
            longest = self.getPalindrome(s, mid, mid + 1)
            ans = longest if len(longest) > len(ans) else ans
        return ans
    
    def getPalindrome(self, s, left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return s[left + 1: right]

File: 5. Longest Palindromic Substring/longestPalindrome.java
class Solution {
    public String longestPalindrome(String s) {
        // write your code here
        if (s.length() <= 1) {
            return s;
        }
        
        int n = s.length();
        boolean[][] isPalindrome = new boolean[n][n];
        for (int i = 0; i < n; i++) {
            isPalindrome[i][i] = true;
            if (i > 0) {
                isPalindrome[i][i - 1] = true;
            }
        }
        
        int start = 0, longest = 1;
        for (int length = 2; length <= n; length++) {
            for (int i = 0; i + length <= n; i++) {
                int j = i + length - 1;
                isPalindrome[i][j] = isPalindrome[i + 1][j - 1] && s.charAt(i) == s.charAt(j);
                if (isPalindrome[i][j] && longest < j - i + 1) {
                    start = i;
                    longest = j - i + 1;
                }
            }
        }
        return s.substring(start, start + longest);
    }
}